### ISSUE TITLE
Memory leak when computing multiple SOAPs in the same script

### ISSUE THREAD

**Issue by benoitba** [SINGROUP/dscribe#69]
Hi,

I am currently working on a project to use SOAP vectors as descriptors of atomic environments for protein pockets.
I was working with another package until now, and would like to use dscribe to compute soaps and derivatives. However working with protein system is quite large, and there is an issue of memory leak that I found while computing soaps for hundreds of proteins.

I tried to show this with a script that compute the same soap for 1000 times, and check the memory usage throughout the iterations. Doing this for water only does show the behaviour, but it is way more visible in terms of memory when doing this for a protein (what I'm actually doing is that I compute the SOAP integrating the protein atom environment but using ligand atomic positions).

![memory_leak_dscribe_soaps](https://user-images.githubusercontent.com/29234619/135878089-6a237d6a-052d-4281-b11b-01b4f7331b20.png)

```python
import os
import psutil
import matplotlib.pyplot as plt
from ase.io import read as aseread
from ase.build import molecule
from dscribe.descriptors import SOAP

# define soap calculator
rcut = 5
nmax = 6
lmax = 4 
sigma = 1
calculator = SOAP(
        rcut = rcut,
        nmax = nmax,
        lmax = lmax,
        sigma = sigma,
        species = "H,C,N,O,S".split(",")
    )

# load molecules to be processed
water = molecule("H2O")

protein_config = aseread('protein_6q48AA.pdb')
protein_config = protein_config[[atom.index for atom in protein_config if atom.symbol in ['C', 'N', 'O', 'S']]]

ligand_config = aseread('ligand_6q48AA.pdb')

# Loop 1 : No action 
memory_usages = []
baseline = psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2
for _ in range(1000) :
    usage = psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2
    memory_usages.append(usage)

memory_usages = [usage - baseline for usage in memory_usages]
plt.plot(memory_usages, label='No action')
    
# Loop 2 : Computing SOAP for a single water system
memory_usages = []
baseline = psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2
for _ in range(1000) :

    soaps = calculator.create(system=water)
    usage = psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2
    memory_usages.append(usage)

memory_usages = [usage - baseline for usage in memory_usages]
plt.plot(memory_usages, label='SOAP for water')
  
# Loop 3 : Computing SOAP for a protein system
memory_usages = []
baseline = psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2
for _ in range(1000) :
        
    soaps = calculator.create(system=protein_config, positions=ligand_config.positions)


    usage = psutil.Process(os.getpid()).memory_info().rss / 1024 ** 2
    memory_usages.append(usage)
    
memory_usages = [usage - baseline for usage in memory_usages]
plt.plot(memory_usages, label='SOAP for protein')

plt.title('Creating SOAPS for a system 1000 times')
plt.xlabel('Iteration')
plt.ylabel('Memory usage (MB)')
plt.legend()
plt.savefig('memory_leak_dscribe_soaps', bbox_inches='tight')
```

Here are the files to reproduce the script if needed. Replace the .txt extension by .pdb
[ligand_6q48AA.txt](https://github.com/SINGROUP/dscribe/files/7279095/ligand_6q48AA.txt)
[protein_6q48AA.txt](https://github.com/SINGROUP/dscribe/files/7279096/protein_6q48AA.txt)

I'm using dscribe v1.1.0, installed with pip.
Do you have any idea where does the memory leak come from ? Is there any work around or ways to fix this ?

Best regards, and thanking you in advance

Benoit

**Comment by lauri-codes** [comment]
Hi @benoitba!

Thank you for reporting this. Indeed it looks like a memory leak based on your test script, thank you for providing it. We will take a look and try to find the culprit. If you want to quickly try something, you can downgrade to some previous version (e.g. 1.0.0), and try if the leak is present there.

-Lauri

**Comment by benoitba** [comment]
The leak is not here on dscribe 1.0.0, so I will work with this version from now on. Thanks a lot for the advice, and hope you find the cause of the leak on the 1.1.0 version :)

**Comment by lauri-codes** [comment]
Now fixed in v1.2.0.

**Comment by GravityPHY** [comment]
Hi, I understand this issue has been fixed. I am curious to know what cause the memory leak in v1.1.0, I tried using del some varibales in the loop and gc.collect(), that way didn't solve the memory leak problem.

**Comment by lauri-codes** [comment]
Hi @GravityPHY!

Yes, the problem has been fixed from v1.2.0 onwards. You may not notice it, but like many other Python libraries that target high performance, we are using C/C++ extensions underneath to do the heavy lifting. This is where the memory leak was happening (missing `free` somewhere). It would be really hard to introduce a memory leak with python only :)
